### ISSUE TITLE
Validation changes break Elasticsearch and Kibana logging

### ISSUE THREAD

**Issue by satnam6502** [GoogleCloudPlatform/kubernetes#10452]
```
$ kubectl create -f es-controller.yaml
The ReplicationController "elasticsearch-logging-v1" is invalid:spec.template.spec.containers[0].ports[1].name: spec.template.spec.containers[0].ports[1].name: invalid value 'es-transport-port': must be a IANA_SVC_NAME (at most 15 characters, matching regex [a-z0-9]([a-z0-9-]*[a-z0-9])* and it must containts at least one letter [a-z], hypens cannot be adjacent to other hyphens): e.g. "http"
```

**Comment by satnam6502** [comment]
I've accidentally added a p1 milestone and can't see how to delete it.

**Comment by satnam6502** [comment]
I think we should file another bug against the validate code which I think is incorrectly rejecting my configs which don't have adjacent hyphens.

**Comment by satnam6502** [comment]
@bprashanth

**Comment by thockin-cc** [comment]
You are over length.  We did change this very late in the game, but it was
that or never change it ever.

**Comment by satnam6502** [comment]
Argh! Well the error message did not help.

**Comment by satnam6502** [comment]
@quinton-hoole can the e2e logging test be promoted out of flaky? This breakage should have been caught by that test.

**Comment by quinton-hoole** [comment]
Sure, I can promote  that test now. That error message is incomprehensible, in addition to being plain wrong. Is there an issue number to fix all this?

**Comment by satnam6502** [comment]
@thockin has corrected me: the error message is "right" in the sense that what I wrote was wrong, because now port names can't be more than 15 characters. I suspect others will run into the same problem so fine-tuning the error message would be good (happy to do it myself -- or leave it to the original author).

**Comment by satnam6502** [comment]
An error message like: "The name must not be longer than 15 characters and it must be composed only of the characters a-z, A-Z or a hyphen which must occur between letters."

**Comment by satnam6502** [comment]
@quinton-hoole also not the e2e logging test should not be promoted until this PR merges.

**Comment by thockin** [comment]
We have sort of resisted breaking errors down further, but maybe we
should.  Would it be useful if this had said:

"invalid value 'es-transport-port': too long.  Value must be a
IANA_SVC_NAME (at most 15 characters, matching regex
[a-z0-9]([a-z0-9-]*[a-z0-9])* and it must containts at least one letter
[a-z], hypens cannot be adjacent to other hyphens): e.g. "http"'

note the extra hint "too long".  We could probably do this for a lot of
error messages.  It might mean breaking regexes into distinct property
tests, but if it is more comprehensible....

**Comment by thockin** [comment]
@bgrant0607 is it worth the risk to try to make API errors more friendly?
It will be logic changes..

**Comment by quinton-hoole** [comment]
Oh, to clarify my hastily written "incomprehensible" comment above (sorry 'bout that):

Current text:

```
The ReplicationController "elasticsearch-logging-v1" is invalid:spec.template.spec.containers[0].ports[1].name: spec.template.spec.containers[0].ports[1].name: invalid value 'es-transport-port': must be a IANA_SVC_NAME (at most 15 characters, matching regex [a-z0-9]([a-z0-9-]*[a-z0-9])* and it must containts at least one letter [a-z], hypens cannot be adjacent to other hyphens): e.g. "http"
```

Things I find confusing about the above (pretending to be normal K8s user):
 
```
spec.template.spec.containers[0].ports[1].name: spec.template.spec.containers[0].ports[1].name
```
Why is the above printed twice?

```
IANA_SVC_NAME
```
I have no idea what the above means, it is not documented anywhere (based on my Google search).  It only exists in the K8s code base.

```
at most 15 characters, matching regex [a-z0-9]([a-z0-9-]*[a-z0-9])* and it must containts at least one letter [a-z], hypens cannot be adjacent to other hyphens
```
Please either give me a proper regex that encapsulates all of the above, not a regex, plus a bunch of other confusing rules.  And/or give me a link to an accurate definition of the valid format.  And please fix the typo "it must containts...".

```
 e.g. "http"
```
It seems like we need more than one example, ideally one that includes hyphens and digits.  And perhaps a counter-example (e.g. "foo-bar but not foo--bar, a-much-too-long-name ..." etc)

**Comment by quinton-hoole** [comment]
PS: As discussed @thockin I don't think that it's necessary/worth the risk to break error messages down into sub-cases for v1.0.  

But it would be great to fix this problem generally after v1.0.

**Comment by thockin** [comment]
Short fix: don't double print field names, make this error more helpful.  I can take this one.

Long-term fix: Add a clue field to ValidationError which indicates what actually went wrong.  Break down all validations into clue-able steps.  Make error message more helpful overall.
